src/tms_service.ts

```diff
@@ -0,0 +1,337 @@
+      );
+      return {
+        ...tileJson,
+        ...{ tiles },
+      };
+    } else {
+      return;
+    }
+  }
+
+  async getUrlTemplate(): Promise<string> {
+    const tileJson = await this._getRasterStyleJson();
+    if (tileJson) {
+      const directUrl = this._proxyPath + this._getAbsoluteUrl(tileJson.tiles[0]);
+      return this._emsClient.extendUrlWithParams(directUrl);
+    } else {
+      return '';
+    }
+  }
+
+  async getUrlTemplateForVector(sourceId: string): Promise<string> {
+    const tileJson = await this._getVectorStyleJsonInlined();
+    if (!tileJson) {
+      return '';
+    }
+    if (tileJson.sources[sourceId] && tileJson.sources[sourceId].tiles) {
+      const directUrl = this._proxyPath + this._getAbsoluteUrl(tileJson.sources[sourceId].tiles[0]);
+      return this._emsClient.extendUrlWithParams(directUrl);
+    } else {
+      return '';
+    }
+  }
+
+  async getVectorStyleSheet(): Promise<EmsVectorStyle | undefined> {
+    return await this._getVectorStyleJsonInlined();
+  }
+
+  async getVectorStyleSheetRaw(): Promise<EmsVectorStyle | undefined> {
+    return await this._getVectorStyleJsonRaw();
+  }
+
+  async getSpriteSheetMeta(
+    isRetina: boolean = false
+  ): Promise<{ png: string; json: EmsSpriteSheet } | undefined> {
+    const metaUrl = await this.getSpriteSheetJsonPath(isRetina);
+    const spritePngs = await this.getSpriteSheetPngPath(isRetina);
+    if (metaUrl && spritePngs) {
+      const metaUrlExtended = this._emsClient.extendUrlWithParams(metaUrl);
+      const jsonMeta = await this._emsClient.getManifest(metaUrlExtended);
+      return {
+        png: spritePngs,
+        json: jsonMeta,
+      };
+    } else {
+      return;
+    }
+  }
+
+  async _getSpriteSheetRootPath(): Promise<string> {
+    const vectorStyleJson = await this._getVectorStyleJsonRaw();
+    if (vectorStyleJson) {
+      return this._proxyPath + this._getAbsoluteUrl(vectorStyleJson.sprite);
+    } else {
+      return '';
+    }
+  }
+
+  async _getUrlTemplateForGlyphs(): Promise<string> {
+    const vectorStyleJson = await this._getVectorStyleJsonRaw();
+    if (vectorStyleJson) {
+      return this._proxyPath + this._getAbsoluteUrl(vectorStyleJson.glyphs);
+    } else {
+      return '';
+    }
+  }
+
+  async getSpriteSheetJsonPath(isRetina: boolean = false): Promise<string> {
+    const spriteSheetRootPath = await this._getSpriteSheetRootPath();
+    if (spriteSheetRootPath) {
+      const suffix = isRetina ? '@2x' : '';
+      return spriteSheetRootPath + suffix + '.json';
+    } else {
+      return '';
+    }
+  }
+
+  async getSpriteSheetPngPath(isRetina: boolean = false): Promise<string> {
+    const spriteSheetRootPath = await this._getSpriteSheetRootPath();
+    if (spriteSheetRootPath) {
+      const suffix = isRetina ? '@2x' : '';
+      return spriteSheetRootPath + suffix + '.png';
+    } else {
+      return '';
+    }
+  }
+
+  getDisplayName(): string {
```

Comment:
There is some duplication in this class and `TMSService` that potentially could be derived to an abstract parent class to contain this method, and those below about attributions and identifier methods.